Improve error checking of Storage._writeFile and fix serialization flakiness

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46036 Improve error checking of Storage._writeFile.**

Previously, this function didn't do error-bounds checking on the GetItem (GET_ITEM) calls, which led to issues like https://github.com/pytorch/pytorch/issues/46020.

A better solution would be to use pybind, but given writing the file is going to dominate bounds checking, this is strictly better.

Differential Revision: [D24228370](https://our.internmc.facebook.com/intern/diff/D24228370)

gchanan committed:
Improve error checking of Storage._writeFile.

Previously, this function didn't do error-bounds checking on the GetItem (GET_ITEM) calls, which lead to issues like https://github.com/pytorch/pytorch/issues/46020.

A better solution would be to use pybind, but given writing the file is going to dominate bounds checking, this is strictly better.

[ghstack-poisoned]

t-vi commented:
If you want CI to pass on this, add a `False` third parameter to the call to `_write_file`  in `_save` in serialization.py. This could then be cherry-picked on the release branch.

gchanan committed:
Update on "Improve error checking of Storage._writeFile."


Previously, this function didn't do error-bounds checking on the GetItem (GET_ITEM) calls, which lead to issues like https://github.com/pytorch/pytorch/issues/46020.

A better solution would be to use pybind, but given writing the file is going to dominate bounds checking, this is strictly better.

[ghstack-poisoned]

gchanan committed:
Update on "Improve error checking of Storage._writeFile."


Previously, this function didn't do error-bounds checking on the GetItem (GET_ITEM) calls, which lead to issues like https://github.com/pytorch/pytorch/issues/46020.

A better solution would be to use pybind, but given writing the file is going to dominate bounds checking, this is strictly better.

[ghstack-poisoned]

mruberry commented on test/test_serialization.py:
```diff
@@ -747,6 +748,7 @@ def run(self, *args, **kwargs):
         with serialization_method(use_zip=True):
             return super(TestSerialization, self).run(*args, **kwargs)
 
+instantiate_device_type_tests(TestBothSerialization, globals())
```
I'm surprised this didn't cause failures with the inherited tests from the mixin.

mruberry commented on torch/csrc/generic/StorageMethods.cpp:
```diff
@@ -221,9 +221,9 @@ static PyObject * THPStorage_(fromFile)(PyObject *_unused, PyObject *args, PyObj
-  PyObject *file = PyTuple_GET_ITEM(args, 0);
-  bool is_real_file = PyTuple_GET_ITEM(args, 1) == Py_True;
-  bool save_size = PyTuple_GET_ITEM(args, 2) == Py_True;
+  PyObject *file = PyTuple_GetItem(args, 0);
```
Bounds checking is cool.

Reviewer's note: https://docs.python.org/3/c-api/tuple.html#c.PyTuple_GetItem

mruberry commented on torch/serialization.py:
```diff
@@ -488,7 +488,7 @@ def persistent_id(obj):
             # Copy to a buffer, then serialize that
             buf = io.BytesIO()
-            storage._write_file(buf, _should_read_directly(buf))
+            storage._write_file(buf, _should_read_directly(buf), False)
```
Let's hope no one else is using _write_file.

mruberry commented:
Technically a test could be added for _write_file throwing an indexerror.

mruberry commented on test/test_serialization.py:
```diff
@@ -585,8 +586,8 @@ def __exit__(self, *args, **kwargs):
 
 class TestBothSerialization(TestCase, SerializationMixin):
```
This will run the tests in the mixin an additional time with the same arguments since they're now instantiated on CUDA and CPU:

test_load_error_msg (__main__.TestBothSerializationCPU) ... ok (0.102s)
test_load_error_msg (__main__.TestBothSerializationCUDA) ... ok (0.045s)

These tests run so quickly it's OK, but I think it'd also be fine to remove the mixin here. The next two test classes run the mixin tests with _use_new_zipfile_serialization=False/True, anyway, so this is just testing its default value (redundantly, since the default value is True).

gchanan committed:
Update on "Improve error checking of Storage._writeFile."


Previously, this function didn't do error-bounds checking on the GetItem (GET_ITEM) calls, which led to issues like https://github.com/pytorch/pytorch/issues/46020.

A better solution would be to use pybind, but given writing the file is going to dominate bounds checking, this is strictly better.

[ghstack-poisoned]
